### DISCUSSION TITLE
[Archive] Warning when event functions are defined and there is no draw()

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#93] [Ideas]
**[UPDATE: This idea turned out to be unnecessary due to other fixes/behavioral changes in py5!]**

**py5** already has a number of very useful error messages, like when you are trying to import py5 but the sketch is being executed by the sketch runner for imported mode .

I wonder if it would be a good idea to have a console warning when you define any of the event functions (like `mouse_pressed()` or `key_pressed()`)  but you forget to define a `draw()` function (which provides the main loop, even when empty).

I would not want this to raise an exception, to abort execution, as I have left unused event functions and used `setup()` only sketches as exploratory/debugging steps (commenting out `draw()`) during development. Just a message like "It seems like you might have defined one or more event functions.... Beware you need to define `draw()` for your sketch to be able to invoke them."

Do you think this is overkill? I think I learned about this a long time a go, maybe in some "interaction" tutorial.

**Comment by hx2A** [comment]
This is a good idea, and might potentially alleviate some confusion for someone someday. I'll use the Python warnings library to do it. This is an easy thing to add.

**Comment by hx2A** [comment]
@villares so I coded these warning messages and tested it and discovered that `mouse_clicked()` and `key_pressed()` events do in fact get called in a sketch that has no `draw()` method. Ha! Not what we expected. This is the case for the default renderer and P2D. I am working on Linux.

It is possible this is a side effect of my fix for #89 & #82, so you might see something different when you test on your machine. I don't think it is a problem that they do get called, I think it is actually an improvement. Do you agree? And if so, that there is no need to add a warning message?

**Reply by villares** [reply]
Wow, that's curious! So they are allowed to update the drawing even without an explicit `draw()`!
Only some of the events, like, `mouse_clicked()`, but not `mouse_pressed()`? `key_pressed()`, but not `key_released()`?

I agree, no need for the warning then....  

I'm mostly on Linux too, so when I get the chance I'll try it on windows... And I have to check I have the most recent build.

**Reply by hx2A** [reply]
> It is possible this is a side effect of my fix for #89 & #82, so you might see something different when you test on your machine.

My hunch was correct. If I used the latest release, the user event functions do not get called if there is no `draw()` method. If I use the development build, which contains the fix for #89 and #82, the user event functions do get called.

There's no longer a need for a warning message and I'll close the issue I opened for this change.
